### PR TITLE
[SYCL][Graph] Explicit update with index only

### PR DESCRIPTION
**Pull request by reble**
This patch removes the match by value functionality,
thus an index is always required to update a kernel parameter.

### CODE REVIEW

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 1]
```diff
@@ -354,6 +359,87 @@ struct graphs_support;
...
+template <>
+class dynamic_parameter<dynamic_parameter_type::usm> {
+public:
+  dynamic_parameter(command_graph<graph_state::modifiable> graph,
+                    std::string label = "");
+
+  void register(void* param, node node);
+
+  void register(void* param);
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 2]
```diff
@@ -354,6 +359,87 @@ struct graphs_support;
...
+class dynamic_parameter<dynamic_parameter_type::accessor> {
+public:
+  dynamic_parameter(command_graph<graph_state::modifiable> graph,
+                    std::string label = "");
+
+  template <typename T>
+  void register(T param, node node);
+
+  template <typename T>
+  void register(T param);
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 3]
```diff
@@ -460,12 +550,425 @@ edges.
...
+any of the dynamic parameters on an executable graph is the new node argument
+on the next submissions of that executable graph.
+
+Registration is done either using the argument index (known when argument
+originally set using `set_arg()`/`set_args()`), or by matching the argument
+value. Matching for USM typed dynamic parameters is done by direct comparison
+of the pointer values. Matching for accessors is done by verifying the
+underlying managed data is the same. Raw and scalar typed dynamic parameters
+can't be registered by matching however, as the chances of unintentional
+collisions in the comparison are too high.
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
Registration is done by using the argument index (known when argument
originally set using `set_arg()`/`set_args()`).
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 4]
```diff
@@ -460,12 +550,425 @@ edges.
...
+dynamic_parameter<usm> DynParam(Graph);
+DynParam.register(PtrA, Node);
+DynParam.update(PtrB, ExecGraph);
+
+// Exception thrown here as no match against original value.
+DynParam.register(PtrB, Node);
+
+// No-op as dynamic parameter already registered with 'PtrA'.
+DynParam.register(PtrA, Node);
+----
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 5]
```diff
@@ -460,12 +550,425 @@ edges.
...
+underlying managed data is the same. Raw and scalar typed dynamic parameters
+can't be registered by matching however, as the chances of unintentional
+collisions in the comparison are too high.
+
+When registration is done by matching the node argument value, it must
+be the original argument value. Attempting to register the dynamic parameter
+again to the same node argument by matching against an updated value will not
+work. This is because registration operates on a modifiable graph, while update
+operates on an executable graph. For example,
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 6]
```diff
@@ -460,12 +550,425 @@ edges.
...
+Parameters:
+
+* `param` - Accessor used as a node argument.
+
+Exceptions:
+
+* Throws with error code `invalid` if no node arguments are matched to `param`
+  in the graph.
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 7]
```diff
@@ -460,12 +550,425 @@ edges.
...
+Parameters:
+
+* `param` - USM pointer used as a node argument.
+
+Exceptions:
+
+* Throws with error code `invalid` if no node arguments are matched to `param`
+  in the graph.
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 8]
```diff
@@ -460,12 +550,425 @@ edges.
...
+Exceptions:
+
+* Throws with error code `invalid` if `param` does not match any arguments in
+  `node`.
+* Throws with error code `invalid` if the type of `node` is not a kernel
+  execution.
+* Throws with error code `invalid` if `node` is not a member of the graph that
+  this dynamic parameter was created with.
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 9]
```diff
@@ -544,6 +1047,80 @@ graph LR
...
+The aspects of a kernel execution node that can be configured during update are:
+
+* Parameters to the kernel.
+* ND-Range of the kernel.
+
+Two methods are provided by the API to the user for performing this update.
+An API for updating the whole graph object, which is most useful when the
+graph was recorded, and an individual node update API.
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 10]
```diff
@@ -544,6 +1047,80 @@ graph LR
...
+possible.
+
+The `property::graph::updatable` property also allows an executable graph to be
+submitted for execution while a previous submission of the same executable
+graph instance is still executing. This is because the ability to change the
+graph inputs/outputs can remove the data race conditions that could otherwise
+exist if the same executable graph was executed concurrently.
+
+===== Whole Graph Update
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 11]
```diff
@@ -544,6 +1047,80 @@ graph LR
...
+`dynamic_parameter` with a node using `dynamic_parameter::register()`, the
+dynamic parameter is matched to a node argument, and the `dynamic_parameter`
+can then be used to update that node argument in future.
+
+If a user doesn't have the individual `node` handles to register parameters
+with, and the dynamic parameter type isn't raw or scalar, they can use the
+`dynamic_parameter::register()` API that doesn't take a `node` argument.
+This entry-point iterates through the modifiable graph nodes, and tries to
+register the parameter against every kernel node in the graph.
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 12]
```diff
@@ -544,6 +1047,80 @@ graph LR
...
+===== Individual Node Update
+
+Memory parameters to individual nodes in a graph in the `executable` state
+can be updated between graph executions using dynamic parameters. When a
+`dynamic_parameter` instance is created with a modifiable state graph it is
+empty, with no associated kernel arguments. By registering the
+`dynamic_parameter` with a node using `dynamic_parameter::register()`, the
+dynamic parameter is matched to a node argument, and the `dynamic_parameter`
+can then be used to update that node argument in future.
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
empty, with no associated kernel arguments. The `dynamic_parameter`
can be used to update that node argument in the future.
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 13]
```diff
@@ -1518,6 +2154,97 @@ submitted in its entirety for execution via
...
+      const size_t i = it[0];
+      ptrY[i] += ptrX[i];
+    });
+   },
+   {sycl_ext::property::node::depends_on(nodeA)});
+
+// Register nodeB dynamic parameters
+dynParamInput.register(ptrX, nodeB);
+dynParamInput.register(ptrY, nodeB);
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 14]
```diff
@@ -1518,6 +2154,97 @@ submitted in its entirety for execution via
...
+// Change ptrX argument to node A & B to ptrZ
+dynParamInput.update(ptrZ, execGraph);
+
+// Change myScalar argument to node A to newScalar
+int newScalar = 12;
+dynParamScalar.update(newScalar, execGraph);
+
+// Change ptrY argument to node B to ptrQ
+dynParamOutput.update(ptrQ, execGraph);
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 15]
```diff
@@ -460,12 +550,425 @@ edges.
...
+
+Exceptions:
+
+* Throws with error code `invalid` if `param` does not match any arguments in
+  `node`.
+* Throws with error code `invalid` if `param` is not an accessor.
+* Throws with error code `invalid` if the type of `node` is not a kernel
+  execution.
+* Throws with error code `invalid` if `node` is not a member of the graph that
+  this dynamic parameter was created with.
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 16]
```diff
@@ -776,9 +1195,66 @@ Exceptions:
...
+were added.
+
+|
+[source,c++]
+----
+std::vector<node> get_root_nodes() const;
+----
+|Returns a list of all nodes in the graph which have no dependencies in the
+order they were added to the graph.
+
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 17]
```diff
@@ -1518,6 +1995,81 @@ submitted in its entirety for execution via
   ...
 ----
 
+=== Dynamic Parameter Update
+
+Example showing a graph with two kernel nodes, the first created using a kernel
+bundle with `handler::set_args()`, having its node arguments updated.
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
Example showing a graph with a single kernel node that is created using a kernel
bundle with `handler::set_args()` and having its node arguments updated.
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 18]
```diff
@@ -1518,6 +1995,81 @@ submitted in its entirety for execution via
...
+// Create graph dynamic parameters
+dynamic_parameter<sycl_ext::dynamic_parameter_type::usm> dynParamInput(myGraph);
+dynamic_parameter<sycl_ext::dynamic_parameter_type::usm> dynParamOutput(myGraph);
+dynamic_parameter<sycl_ext::dynamic_parameter_type::scalar> dynParamScalar(myGraph);
+
+// First node uses ptrX as an input & output parameter, with operand
+// mySclar as another argument.
+int myScalar = 42;
+node nodeA = myGraph.add([&](handler& cgh) {
+    cgh.set_args(ptrX, myScalar);
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
    cgh.set_args(ptrX, ptrY, myScalar);
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 19]
```diff
@@ -1518,6 +1995,81 @@ submitted in its entirety for execution via
...
+    cgh.parallel_for(range {n}, builtinKernel);
+});
+
+// Register nodeA dynamic parameters
+dynParamInput.register(0, nodeA);  // Argument index 0 is ptrX
+dynParamInput.register(1, nodeA);  // Argument index 1 is myScalar
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
dynParamInput.register(2, nodeA);  // Argument index 1 is myScalar
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 20]
```diff
@@ -1518,6 +1995,81 @@ submitted in its entirety for execution via
...
+
+// Create an executable graph with the updatable property.
+auto execGraph = myGraph.finalize({sycl_ext::property::graph::updatable});
+
+// Execute graph, then update without needing to wait for it to complete
+myQueue.ext_oneapi_graph(execGraph);
+
+// Change ptrX argument to node A & B to ptrZ
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
// Change ptrX argument to node A to ptrZ
```

**Comment by EwanC** [comment]
1) I think we need to change how the `dynamic_parameter` class is templated to as suggested by Greg in https://github.com/reble/llvm/pull/340#discussion_r1428505807

```
template<typename T>
class dynamic_parameter {
 public:
  dynamic_parameter(const T& param, command_graph<graph_state::modifiable> graph, std::string label = "");
  register(const handler &cgh);
  update(const T& newval, command_graph<graph_state::executable> execGraph);
};
```
2) I think we need to remove wording about changing concurrent execution management and revisit this later. We haven't fully thought this through and the implementation change will drag out the time to implement this spec change
```
If `graph` is submitted multiple times and
`property::graph::updatable_graph` is not set, dependencies are automatically
added by the runtime to prevent concurrent executions of an identical graph.
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 21]
```diff
@@ -355,6 +361,17 @@ struct graphs_support;
...
 } // namespace info
 
+template <typename ValueT>
+class dynamic_parameter {
+public:
+  dynamic_parameter(command_graph<graph_state::modifiable> graph);
+
+  void register(int argIndex, const handler& cgh);
+
+  void update(const T& newValue,
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
  void update(const ValueT& newValue,
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 22]
```diff
@@ -471,6 +492,110 @@ namespace sycl::ext::oneapi::experimental {
...
+  for the resulting node.
+* Throws with error code `invalid` if the type of the command-group is not a
+  kernel execution.
+* Throws with error code `invalid` if the resulting node is not associated with
+  the graph that this dynamic parameter was created with.
+
+|
+[source,c++]
+----
+void update(ValueT newValue,
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: `valueT newValue` is `const ValueT& newValue` in the declaration earlier in the document

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 23]
```diff
@@ -471,6 +492,110 @@ namespace sycl::ext::oneapi::experimental {
...
+
+|Updates parameters in `execGraph` nodes associated with the dynamic
+parameter to `newValue`. This new value will not affect any prior submissions
+of `execGraph` and will take affect only for future submissions. It is not an
+error if `newValue` is set to the current parameter value in nodes of
+`execGraph`, i.e. no update occurred.
+
+Parameters:
+
+* `newValue` - Value to the parameters being updated to.
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: This description doesn't read that nicely, could be something like.
```suggestion
* `newValue` - Value to update the registered parameters to.
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 24]
```diff
@@ -471,6 +492,110 @@ namespace sycl::ext::oneapi::experimental {
...
+* `execGraph` - The executable graph to update the node parameters for.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if
+  `property::graph::updatable` was not set when `execGraph` was created.
+* Throws synchronously with error code `invalid` if `newValue` is not a pointer
+  to a valid USM allocation.
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: This error code seems outdated

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 25]
```diff
@@ -548,6 +673,66 @@ graph LR
...
+have been set when the graph was created during finalization. Otherwise, an
+exception will be thrown if a user tries to update an executable graph. This
+guarantee allows the backend to provide a more optimized implementation, if
+possible.
+
+The `property::graph::updatable` property also allows an executable graph to be
+submitted for execution while a previous submission of the same executable
+graph instance is still executing. This is because the ability to change the
+graph inputs/outputs can remove the data race conditions that could otherwise
+exist if the same executable graph was executed concurrently.
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: This we decided to defer this functionality, could be another item in the Issues section

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 26]
```diff
@@ -615,6 +800,9 @@ associated with `syclContext`.
 * Throws synchronously with error code `invalid` if `syclDevice`
   <<device-info-query, reports this extension as unsupported>>.
 
+* Throws synchronously with error code `invalid` if the backend associated
+with `syclDevice` is not supported.
+
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: New wording looks unrelated to this PR

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 27]
```diff
@@ -1548,6 +1780,81 @@ submitted in its entirety for execution via
...
+dynamic_parameter<sycl_ext::dynamic_parameter_type::scalar> dynParamScalar(myGraph);
+
+// First node uses ptrX as an input & output parameter, with operand
+// mySclar as another argument.
+int myScalar = 42;
+node nodeA = myGraph.add([&](handler& cgh) {
+    cgh.set_args(ptrX, ptrY, myScalar);
+    // Register nodeA dynamic parameters
+    dynParamInput.register(0, cgh);  // Argument index 0 is ptrX
+    dynParamInput.register(2, cgh);  // Argument index 1 is myScalar
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: * Comments says index 1 but we're passing index 2
* should this also be `dynParamScalar`

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 28]
```diff
@@ -1548,6 +1780,81 @@ submitted in its entirety for execution via
...
+kernel_bundle<bundle_state::executable> myBundle =
+      get_kernel_bundle(myContext, { myDevice }, builtinKernelIds);
+kernel builtinKernel = myBundle.get_kernel(builtinKernelIds[0]);
+
+// Graph containing a two kernels node
+sycl_ext::command_graph myGraph(myContext, myDevice);
+
+// Create graph dynamic parameters
+dynamic_parameter<sycl_ext::dynamic_parameter_type::usm> dynParamInput(myGraph);
+dynamic_parameter<sycl_ext::dynamic_parameter_type::usm> dynParamOutput(myGraph);
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: This is never used in this example

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 29]
```diff
@@ -1752,6 +2059,36 @@ runtime.
...
+
+The current approach for SYCL-Graph is more complex, in that:
+
+- Updating arguments is handled exclusively through the `dynamic_parameter`
+  class, which both handles updating values and updating the graph.
+- ND-ranges for kernels are updated directly through the executable graph,
+  without first updating the node separately.
+
+Having a more unified approach to updating nodes/graphs may provide a more
+concise and easy to use API.
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Been thinking about this since we chatted yesterday, and I'm not mad about going forward with an API that we think we could do better with and will change very soon after delivering. If it'll only take a day or two to modify this so that `nd_range` is updated on the node, and nodes are passed to a graph to update, then it might be worth doing.

**Comment by Bensuo** [comment]
I've pushed a commit wihch proposes some API changes to (IMO) improve the usage of the API, and make things a bit clearer in what's happening.

The main change is the ND-ranges are updated on the Node itself, and that updates to the executable graph are triggered by explicitly passing updates nodes to the graph. This to me serves two purposes:

- Unifies how Node param and ND-range updates are applied to the executable graph
- Makes explicit for the user where the update actually occurs. Previously `dynamic_parameter` had a lot of responsibility, both updating nodes and also applying those updates to the executable graph. IMO this is clearer and easier to use.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 30]
```diff
@@ -355,7 +361,20 @@ struct graphs_support;
...
+public:
+  dynamic_parameter(command_graph<graph_state::modifiable> graph);
+
+  void register(int argIndex, const handler& cgh);
+
+  void update_value(const ValueT& newValue);
+};
+
+class node {
+  template <int Dimensions>
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: don't know if we want a `node() = delete;` here

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 31]
```diff
@@ -467,10 +489,131 @@ The `node` class provides the {crs}[common reference semantics].
...
+
+Parameters:
+
+* `executionRange` - The new value for the ND-Range.
+
+Exceptions:
+
+* Throws with error code `nd_range` if `executionRange` is an invalid ND-Range
+  for this node.
+
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Maybe this too
```
*  Throws with error code `invalid` if the type of the node is not a kernel execution.
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 32]
```diff
@@ -467,10 +489,131 @@ The `node` class provides the {crs}[common reference semantics].
...
+  for the resulting node.
+* Throws with error code `invalid` if the type of the command-group is not a
+  kernel execution.
+* Throws with error code `invalid` if the resulting node is not associated with
+  the graph that this dynamic parameter was created with.
+
+|
+[source,c++]
+----
+void update(const ValueT &newValue);
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: called ` update_value` above with left aligned `&`

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 33]
```diff
@@ -548,6 +691,64 @@ graph LR
...
+Parameter updates are performed using a `dynamic_parameter` instance by
+calling `dynamic_parameter::update()` to update all the parameters of nodes to
+which the `dynamic_parameter` is registered.
+
+The other node configuration that can be updated is the ND-Range, this can
+be set through `node::update_nd_range()` but does not require any prior
+registration.
+
+These updated nodes can then be passed to 
+`command_graph<graph_state::executable>::update()` will will update the
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
`command_graph<graph_state::executable>::update()` which will update the
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 34]
```diff
@@ -808,7 +1010,70 @@ the file extension is not supported or if the write operation failed.
...
+|Member function|Description
+
+|
+[source,c++]
+----
+void update(node& node);
+----
+
+| Updates an executable graph node that corresponds to `node`. `node` must be a
+kernel exewcution node. Argument values which have associated dynamic parameters
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
kernel execution node. Argument values which have associated dynamic parameters
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 35]
```diff
@@ -808,7 +1010,70 @@ the file extension is not supported or if the write operation failed.
...
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if
+  `property::graph::updatable` was not set when the executable graph was
+  created.
+* Throws with error code `invalid` if `node` is not part of the
+  graph.
+
+  |
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: extra indent on `|` that could be removed

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc (hunk 29 above)
Comment (EwanC) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: new commit looks good, think we can remove this issue item

**Review by EwanC — APPROVED** [review]
Comments are minor nits, LGTM

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc (hunk 31 above)
Comment (Bensuo) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Good catch, added this.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 36]
```diff
@@ -467,10 +492,134 @@ The `node` class provides the {crs}[common reference semantics].
...
+  execution.
+
+|===
+
+==== Dynamic Parameters
+
+Dynamic parameters are arguments to a node's command-group which can be updated
+by the user after the node has been added to a graph. Updating the value of a
+dynamic parameter will be reflected in the modifiable graph which contains this
+node. Executable graphs can then be updated with these update nodes
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
node.
```
Sentence doesn't make sense to me. Maybe we can be more specific or just remove it.

Comment (Bensuo) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: I've tried to rework this sentence to make it read better. I think it's important to keep something here that explains how the update makes it's way to the executable graph in the end though.

**Comment by Bensuo** [comment]
Closing in favour of upstream PR: https://github.com/intel/llvm/pull/12486
